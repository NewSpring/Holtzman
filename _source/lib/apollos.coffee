###

  Apollos.name

  @example get a string with the name of this system

    console.log Apollos.name

###
Apollos.name = "Apollos"


###

  Apollos.user

  @example get the currently logged in user

    console.log Apollos.user()._id

###
Apollos.user = (id) ->

  if Meteor.isServer
    return {}

  if Meteor.isClient
    user = Meteor.user()

  return user or {}


Apollos.user.login = {}

if Meteor.isClient
  Meteor.subscribe "currentUserPerson"
else
  Meteor.publish "currentUserPerson", ->
    user = Apollos.users.findOne @.userId
    query = {}
    queryOk = false

    if not user
      return

    if user.personGuid
      query.guid = user.personGuid
      queryOk = true

    # we need to move the rock stuff out of this package
    if user.rock and user.rock.personId
      query.personId = user.rock.personId
      queryOk = true

    if not queryOk
      return

    return Apollos.people.find query


###

  Apollos.person

  @example get the currently logged in user's person document

    console.log "Hello, #{Apollos.person().firstName}"

###
Apollos.person = (user) ->

  if Meteor.isServer
    if not user?._id
      throw new Meteor.Error ("user is requried for server lookup")
      return

  if Meteor.isClient
    user or= Apollos.user()

  if user.personGuid
    person = Apollos.people.findOne
      guid: user.personGuid


  # we shouldnt have any code for rock in the core package
  # this will probably need to adjust the rock package

  # if not person and user.rock and user.rock.personId
  #   person = Apollos.people.findOne
  #     personId: user.rock.personId

  return person or {}






if Meteor.server
  Accounts.onCreateUser (options, user) ->

    if options.profile
      user.profile = options.profile

    if user.profile?.guest is true
      return user

    person = Apollos.person user

    # # no existing user so create one
    if not Object.keys(person).length
      if not user.personGuid
        user.personGuid = Apollos.utilities.makeNewGuid()

      Apollos.people.upsert({guid: user.personGuid}, {
        $set:
          preferredEmail: user.emails[0].address
      })

    return user

###

  Apollos.user.create

  @example create a new user

    Apollos.user.create "xyz@abc.cc", "password1", (error) ->
      if error
        console.log error

  @param email is the email address for the user to use to login with
  @param password is the plain-text password the user will authenticate with
  @param callback is the function that will be called with an error if so

###
Apollos.user.create = (email, password, callback) ->

  Accounts.createUser
    email: email
    password: password
  ,
    callback

###

  Apollos.user.forgotPassword

  @example request a forgot password email

    Apollos.user.forgotPassword "xyz@abc.cc", (error) ->
      if error
        console.log error

  @param email is the email address for the user to use to login with
  @param callback is the function that will be called with an error if so

###
Apollos.user.forgotPassword = (email, callback) ->

  return Accounts.forgotPassword
    email: email
  ,
    callback

###

  Apollos.user.resetPassword

  @example changes password using reset token

    Apollos.user.resetPassword token, newPassword, (error) ->
      if error
        console.log error

  @param token is the reset token emailed to the user
  @param newPassword is what to change the password to
  @param callback is the function that will be called with an error if so

###
Apollos.user.resetPassword = (token, newPassword, callback) ->

  return Accounts.resetPassword token, newPassword, callback


###

  Apollos.user.changePassword

  @example changes password using old password

    Apollos.user.changePassword oldPassword, newPassword, (error) ->
      if error
        console.log error

  @param oldPassword is the user's current password
  @param newPassword is what to change the password to
  @param callback is the function that will be called with an error if so

###
Apollos.user.changePassword = (oldPassword, newPassword, callback) ->

  return Accounts.changePassword oldPassword, newPassword, callback

if Meteor.isServer
  # TODO this is not secure
  Apollos.emailTemplates = Accounts.emailTemplates
