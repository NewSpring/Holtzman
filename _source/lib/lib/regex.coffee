_addRegexTestFunction = (regex) ->
  funcName = "is#{regex.charAt(0).toUpperCase()}#{regex.slice(1)}"
  Apollos.validate[funcName] = (str) ->
    return Apollos.regex[regex].test str


Apollos.regex =

  # @TODO choose which email regex to use
  #
  # this was ours
  # email: /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
  #
  # this is what rock uses:
  email: /[\w\.\'_%-]+(\+[\w-]*)?@([\w-]+\.)+[\w-]+/

  bcrypt: /^\$2a\$10\$[\/\.a-zA-Z0-9]{53}$/

  phoneNumber: /^[1-9]([0-9]{6}|[0-9]{9})$/

  guid: /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

  visa: /^4[0-9]{12}(?:[0-9]{3})?$/

  mastercard: /^5[1-5][0-9]{14}$/

  americanExpress: /^3[47][0-9]{13}$/

  discover: /^6(?:011|5[0-9]{2})[0-9]{12}$/

  startOfVisa: /^4[0-9]{0,15}$/

  startOfMastercard: /^5$|^5[1-5][0-9]{0,14}$/

  startOfAmEx: /^3$|^3[47][0-9]{0,13}$/

  startOfDiscover: /^6$|^6[05]$|^601[1]?$|^65[0-9][0-9]?$|^6(?:011|5[0-9]{2})[0-9]{0,12}$/


Apollos.addRegex = (obj) ->

  for key, value of obj
    Apollos.regex[key] = value

  return


Apollos.validate =

  acceptedCardTypes: ["Visa", "Mastercard", "AmEx", "Discover"]

  guessCardType: (number) ->

    number = number.replace /-/g, ''

    for cardType in Apollos.validate.acceptedCardTypes

      if Apollos.validate["isStartOf#{cardType}"] number
        return cardType

    return false


  getCardType: (number) ->

    number = number.replace /-/g, ''

    for cardType in Apollos.validate.acceptedCardTypes

      if Apollos.validate["is#{cardType}"] number
        return cardType

    return false


  isBcryptHash: (str) ->
    return @.isBcrypt(str)


for regex of Apollos.regex
  _addRegexTestFunction regex
