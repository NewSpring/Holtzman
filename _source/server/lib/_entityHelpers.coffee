Apollos.documentHelpers =


  translate: (singular, data, platform) ->

    if not platform
      Apollos.debug "must specify platform to translate to"
      return

    platform = platform.toUpperCase()

    if not Apollos[singular]._dictionary[platform]
      Apollos.debug "no translation found for #{platform} in #{singular}"
      return

    # translate
    return Apollos[singular]._dictionary[platform] data


  ###

    Apollos.documentHelpers.update

    @example update a doc in apollos with data from a platform

      Apollos.documentHelpers.update "person", "people", person, platform

    @param singular [String] The singular form of the entities name
    @param plural [String] The plural form of the entities name
    @param doc [Object] existing doc from other service to be updated
    @param platform [String] platform to be updated from

  ###
  update: (singular, plural, doc, platform) ->

    doc = Apollos[singular].translate doc, platform

    if not doc
      return

    singularIdKeyValue = {}
    singularIdKeyValue["#{singular}Id"] = doc["#{singular}Id"]

    if platform
      doc.updatedBy = platform.toUpperCase()
    else
      doc.updatedBy = Apollos.name

    query =
      $or: [
        singularIdKeyValue
      ,
        guid: RegExp(doc.guid, "i")
      ]

    matches = Apollos[plural].find query,
      sort:
        updatedDate: 1

    if matches.count() > 1
      # Delete older documents, which are the first ones since they are sorted
      ids = matches.map (m) ->
        return m._id

      ids.pop()

      Apollos[plural].remove
        _id:
          $in: ids

      matches = Apollos[plural].find(query)

    if matches.count() is 1
      existing = matches.fetch()[0]
      mongoId = existing._id

      Apollos[plural].update
        _id: mongoId
      ,
        $set: doc

    else
      mongoId = Apollos[plural].insert doc

    return mongoId


  ###

    Apollos.documentHelpers.delete

    @example take an colection and delete it

      Apollos.documentHelpers.delete("person", "people", 3, plaform)

    @param singular [String] The singular form of the entities name
    @param plural [String] The plural form of the entities name
    @param identifier [Object|String|Number] existing document, _id
    @param platform [String] platform initiating the delete

  ###
  delete: (singular, plural, identifier, platform) ->

    if typeof identifier is "number"
      singularIdKeyValue = {}
      singularIdKeyValue["#{singular}Id"] = identifier
      doc = Apollos[plural].findOne singularIdKeyValue

    else if typeof identifier is "string"
      doc = Apollos[plural].findOne identifier

    else
      doc = identifier

    if typeof doc isnt "object"
      throw new Meteor.Error "Delete Error", "Could not delete #{singular}
        identified by #{identifier}"

    if platform
      doc.updatedBy = platform.toUpperCase()
    else
      doc.updatedBy = Apollos.name

    # We have to update this first so the collection hooks know what to do
    Apollos[plural].update
      _id: doc._id
    ,
      $set:
        "updatedBy": doc.updatedBy

    Apollos[plural].remove doc._id
