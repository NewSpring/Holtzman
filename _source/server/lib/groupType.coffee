Apollos.groupType = {}


###

  Apollos.groupType.translate

  @example take data from a service and format it for Apollos

    Apollos.groupType.translate([obj, platform])

  @param groupType [Object] existing object to be translated
  @param platform [String] platform to be translated to

###
Apollos.groupType.translate = (groupType, platform) ->
  return Apollos.documentHelpers.translate "groupType", groupType, platform


###

  Apollos.groupType.update

  @example update a groupType in apollos with data from another service

    Apollos.groupType.update([obj, platform])

  @param groupType [Object] existing groupType from other service to be updated
  @param platform [String] platform to be update from

###
Apollos.groupType.update = (groupType, platform) ->
  return Apollos.documentHelpers.update "groupType", "groupTypes", groupType, platform


###

  Apollos.groupType.delete

  @example take a groupType and delete it

    Apollos.groupType.delete(user, [platform])

  @param groupType [Object|String|Number] existing document, _id
  @param platform [String] platform initiating the delete

###
Apollos.groupType.delete = (groupType, platform) ->
  Apollos.documentHelpers.delete "groupType", "groupTypes", groupType, platform
