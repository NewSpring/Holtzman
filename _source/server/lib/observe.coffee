changeWasAlreadyHandled = (doc, collection) ->

  if not doc.observationHash
    return false

  hash = hashObject doc

  if doc.observationHash is hash
    return true

  Apollos[collection].update doc._id, $set: observationHash: hash
  doc.observationHash = hash
  return false

hashObject = (obj) ->

  json = JSON.stringify obj
  hash = 0
  i = 0
  len = json.length

  if len is 0
    return hash

  while i < len
    chr = json.charCodeAt(i)
    hash  = ((hash << 5) - hash) + chr
    hash |= 0
    i++

  return hash

Apollos.observe = (collection) ->

  initializing = true

  Apollos[collection].find({},
    fields:
      observationHash: 0
      createdDate: 0
      updatedDate: 0
      updatedBy: 0
  ).observeChanges

    added: (id) ->

      if initializing
        return

      doc = Apollos[collection].findOne id
      if changeWasAlreadyHandled doc, collection
        Apollos.debug "#{collection} #{id} add already handled"
        return

      Apollos.debug "Handling add in #{collection} #{id}"
      Apollos[collection].added or= {}
      platformAddedBy = doc.updatedBy?.toUpperCase()

      for platform, handle of Apollos[collection].added
        if platformAddedBy and (platformAddedBy isnt platform.toUpperCase())
          handle doc

    changed: (id) ->

      doc = Apollos[collection].findOne id
      if changeWasAlreadyHandled doc, collection
        Apollos.debug "#{collection} #{id} change already handled"
        return

      Apollos.debug "Handling change in #{collection} #{id}"
      Apollos[collection].changed or= {}
      platformChangedBy = doc.updatedBy?.toUpperCase()

      for platform, handle of Apollos[collection].changed
        if platformChangedBy and (platformChangedBy isnt platform.toUpperCase())
          handle doc

    removed: (id) ->

      Apollos[collection].deleted or= {}

      for platform, handle of Apollos[collection].deleted
        handle id


  initializing = false



Apollos.observe.remove = (doc, platform, methods) ->

  if not Apollos[doc]
    Apollos.debug "Cannot remove observe methods for #{doc} becuase it doesn't exist"
    return

  if not platform
    Apollos.debug "Must specify platform"
    return

  if not methods
    methods = ["added", "changed", "removed"]

  for method in methods
    delete Apollos[doc][method]


Apollos.observe.add = (doc, platform, methods) ->

  if not Apollos[doc]
    Apollos.debug "Cannot set observe methods for #{doc} becuase it doesn't exist"
    return

  if not platform
    Apollos.debug "Must specify platform"
    return


  for method, handle of methods

    Apollos[doc][method] or= {}

    if Apollos[doc][method][platform]
      Apollos.debug "The #{method} observe handler for #{doc} from #{platform} has already been set"
      return
    Apollos[doc][method][platform] = handle


  return
