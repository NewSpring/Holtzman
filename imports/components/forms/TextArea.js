import { Component, PropTypes } from "react";

import Label from "./components/Label";

export default class TextArea extends Component {

  static propTypes = {
    defaultValue: PropTypes.oneOfType([
      PropTypes.bool,
      PropTypes.string,
    ]),
    status: PropTypes.string,
    disabled: PropTypes.any, // eslint-disable-line
    validation: PropTypes.func,
    errorText: PropTypes.string,
    theme: PropTypes.string,
    type: PropTypes.string,
    error: PropTypes.any, // eslint-disable-line
    classes: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.array,
    ]),
    children: PropTypes.any, // eslint-disable-line
    id: PropTypes.string,
    label: PropTypes.string,
    name: PropTypes.string,
    inputClasses: PropTypes.array, // eslint-disable-line
    hideLabel: PropTypes.bool,
    autofocus: PropTypes.any, // eslint-disable-line
    format: PropTypes.func,
    onChange: PropTypes.func,
    onBlur: PropTypes.func,
    style: PropTypes.object, //eslint-disable-line
    placeholder: PropTypes.string,
    rows: PropTypes.number,
  }

  state = {
    active: false,
    focused: false,
    error: false,
    status: "",
    value: null,
  }

  componentWillMount() {
    if (this.props.defaultValue) {
      this.setState({ active: true });
    }
  }

  componentDidMount() {
    if (this.props.autofocus) {
      this.node.focus();
    }

    // if (this.props.defaultValue) {
    //   this.node.value = `${this.props.defaultValue}`
    // }

    // one day, I dream of a universal browser auto-fill event
    // until then. I'll keep on checking
    const target = this.node;
    this.interval = setInterval(() => {
      if (this._previousValue === target.value || !target.value) { // eslint-disable-line
        return;
      }

      if (!this._previousValue && target.value && !this.state.focused) { // eslint-disable-line
        this.setValue(target.value);
      }

      this._previousValue = target.value; // eslint-disable-line
    }, 20);
  }

  componentWillUpdate(nextProps) {
    if (this.props.defaultValue !== nextProps.defaultValue) {
      this.setValue(nextProps.defaultValue);
      this.setState({ focused: false });
    }
  }

  componentWillUnmount() {
    if (this.interval) {
      clearInterval(this.interval);
    }
  }


  format = (e) => { // eslint-disable-line
    const target = this.node;
    const value = this.node.value;

    if (this.props.format && typeof (this.props.format) === "function") {
      const newValue = this.props.format(value, target, e);
      target.value = newValue;
    }

    if (this.props.onChange && typeof (this.props.onChange) === "function") {
      this.props.onChange(target.value, target, e);
    }
  }

  validate = (e) => {
    const target = this.node;
    const value = target.value;

    if (!value) {
      this.setState({
        active: false,
        error: false,
      });
    }

    this.setState({
      focused: false,
    });

    if (this.props.validation && typeof (this.props.validation) === "function") {
      this.setState({
        error: !this.props.validation(value, target, e),
      });
    }

    if (this.props.onBlur && typeof (this.props.onBlur) === "function") {
      this.props.onBlur(value, target, e);
    }
  }

  focus = () => {
    this.setState({
      active: true,
      error: false,
      focused: true,
    });
  }

  setValue = (value) => {
    const node = this.node;
    node.value = value;
    this.focus();
    this.validate();
  }

  getValue = () => this.node.value;

  setStatus = (message) => {
    this.props.status = message;
  }

  disabled = () => {
    if (this.props.disabled) {
      return this.props.disabled;
    }
    return undefined;
  }

  renderHelpText = () => {
    if ((this.state.error && this.props.errorText) || this.state.status) {
      return (
        <span className="input__status">
          {this.props.errorText || this.state.status}
        </span>
      );
    }
    return undefined;
  }


  style = () => {
    let style = {};

    if (this.props.style) {
      style = { ...style, ...this.props.style };
    }

    if (this.props.disabled) {
      style = {
        ...style,
        ...{
          cursor: "inherit",
        },
      };
    }

    return style;
  }


  render() {
    let inputclasses = [
      "input",
    ];

    // theme overwrite
    if (this.props.theme) { inputclasses = this.props.theme; }
    // state mangaged classes
    if (this.state.active) { inputclasses.push("input--active"); }
    if (this.state.focused) { inputclasses.push("input--focused"); }
    if (this.state.error) { inputclasses.push("input--alert"); }
    // custom added classes
    if (this.props.classes) { inputclasses = inputclasses.concat(this.props.classes); }

    return (
      <div className={inputclasses.join(" ")} style={this.props.style || {}}>
        {(() => {
          if (!this.props.hideLabel) {
            return (
              <Label
                labelFor={
                  this.props.id || this.props.name || this.props.label
                }
                labelName={
                  this.props.label || this.props.name
                }
                disabed={this.disabled()}
              />
            );
          }
          return undefined;
        })()}


        <textarea
          ref={node => (this.node = node)}
          id={this.props.id || this.props.name || this.props.label}
          type={this.props.type}
          placeholder={this.props.placeholder || this.props.label}
          name={this.props.name || this.props.label}
          className={this.props.inputClasses}
          disabled={this.disabled()}
          onBlur={this.validate}
          onFocus={this.focus}
          onChange={this.format}
          style={this.style()}
          rows={this.props.rows || 8}
          defaultValue={this.props.defaultValue}
        />

        {this.renderHelpText()}

      </div>
    );
  }

}
