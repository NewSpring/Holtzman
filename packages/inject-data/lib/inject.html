<script type="text/inject-data"><%= data %></script>
