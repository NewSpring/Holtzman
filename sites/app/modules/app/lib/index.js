

// named collections in EE :(
import Collections from "./collections"

export default {
  Collections
}
