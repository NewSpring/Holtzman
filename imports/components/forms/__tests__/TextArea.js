import { shallow, mount } from "enzyme";
import { shallowToJson } from "enzyme-to-json";
import { reset, startBuffering } from "aphrodite/lib/inject";
import cloneDeep from "lodash.clonedeep";
import TextArea from "../TextArea";

const defaultProps = {
  defaultValue: null,
  status: null,
  disabled: false,
  validation: jest.fn(),
  errorText: null,
  theme: null,
  error: null,
  classes: null,
  id: "2",
  label: "myLabel",
  name: "myName",
  inputClasses: "input classes",
  hideLabel: false,
  autofocus: false,
  onChange: jest.fn(),
  onBlur: jest.fn(),
  style: {
    color: "green",
  },
  placeholder: "myPlaceholder",
  rows: 10,
};

const generateComponent = (additionalProps = {}) => {
  const newProps = {
    ...defaultProps,
    ...additionalProps,
  };
  return <TextArea { ...newProps } />;
};

beforeEach(() => {
  reset();
  startBuffering();
});

afterEach(() => {
  reset();
});

it("renders with props", () => {
  const wrapper = shallow(generateComponent());
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("overrides with theme if present", () => {
  const wrapper = shallow(generateComponent({
    theme: ["override"],
  }));
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("has active state", () => {
  const wrapper = shallow(generateComponent());
  wrapper.setState({ active: true });
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("has focused state", () => {
  const wrapper = shallow(generateComponent());
  wrapper.setState({ focused: true });
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("has error state", () => {
  const wrapper = shallow(generateComponent());
  wrapper.setState({ error: true });
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("appends extra classes", () => {
  const wrapper = shallow(generateComponent({
    classes: ["append", "me"],
  }));
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("can hide the label", () => {
  const wrapper = shallow(generateComponent({
    hideLabel: true,
  }));
  expect(shallowToJson(wrapper)).toMatchSnapshot();
});

it("updates state on mount if default value", () => {
  const wrapper = shallow(generateComponent({
    defaultValue: "2",
  }));
  expect(wrapper.state().active).toBe(true);
});

it("sets value of node on did mount", () => {
  const wrapper = mount(generateComponent({
    defaultValue: "2",
  }));
  expect(wrapper.instance().getValue()).toBe("2");
  expect(wrapper.instance().interval).toBeTruthy();
});

it("updates state on will update", () => {
  const wrapper = mount(generateComponent({
    defaultValue: "2",
  }));
  const nextProps = cloneDeep(defaultProps);
  nextProps.defaultValue = "3";
  wrapper.setState({ focused: true });
  wrapper.instance().componentWillUpdate(nextProps);
  expect(wrapper.state().focused).toBe(false);
});

it("removes interval on unmount", () => {
  const mockClearInterval = jest.fn();
  global.clearInterval = mockClearInterval;
  const wrapper = mount(generateComponent({
    defaultValue: "2",
  }));
  wrapper.instance().componentWillUnmount();
  expect(mockClearInterval).toHaveBeenCalledTimes(1);
});

it("format calls onChange and validation function", () => {
  const mockFormat = jest.fn((x) => x);
  const mockOnChange = jest.fn();
  const mockEvent = {};
  const wrapper = mount(generateComponent({
    defaultValue: "2",
    format: mockFormat,
    onChange: mockOnChange,
  }));
  wrapper.instance().format(mockEvent);
  expect(mockFormat).toHaveBeenCalledTimes(1);
  expect(mockFormat.mock.calls[0][0]).toBe("2");
  expect(mockFormat.mock.calls[0][2]).toBe(mockEvent);
  expect(mockOnChange).toHaveBeenCalledTimes(1);
  expect(mockOnChange.mock.calls[0][0]).toBe("2");
  expect(mockOnChange.mock.calls[0][2]).toBe(mockEvent);
});

it("validate updates state when no value", () => {
  const mockValidation = jest.fn(() => true);
  const wrapper = mount(generateComponent({
    validation: mockValidation,
  }));
  wrapper.setState({ active: true, error: true, focused: true });
  wrapper.instance().validate();
  expect(mockValidation).toHaveBeenCalledTimes(1);
  expect(wrapper.state().active).toBe(false);
  expect(wrapper.state().error).toBe(false);
  expect(wrapper.state().focused).toBe(false);
});

it("validate sets error when invalid", () => {
  const mockValidation = jest.fn(() => false);
  const mockOnBlur = jest.fn();
  const mockEvent = jest.fn();
  const wrapper = mount(generateComponent({
    defaultValue: "1",
    validation: mockValidation,
    onBlur: mockOnBlur,
  }));
  wrapper.instance().validate(mockEvent);
  expect(wrapper.state().error).toBe(true);
  expect(mockOnBlur).toHaveBeenCalledTimes(1);
  expect(mockOnBlur.mock.calls[0][0]).toBe("1");
  expect(mockOnBlur.mock.calls[0][2]).toBe(mockEvent);
});

it("focus changes the state", () => {
  const wrapper = mount(generateComponent());
  wrapper.setState({ error: true });
  wrapper.instance().focus();
  expect(wrapper.state().active).toBe(true);
  expect(wrapper.state().error).toBe(false);
  expect(wrapper.state().focused).toBe(true);
});

it("disabled returns disabled when set", () => {
  const wrapper = shallow(generateComponent({
    disabled: true,
  }));
  const result = wrapper.instance().disabled();
  expect(result).toBe(true);
});

it("disabled returns undefined when not set", () => {
  const wrapper = shallow(generateComponent());
  const result = wrapper.instance().disabled();
  expect(result).toBe(undefined);
});

it("renderHepText returns undefined when no error or status", () => {
  const wrapper = shallow(generateComponent());
  const result = wrapper.instance().renderHelpText();
  expect(result).toBe(undefined);
});

it("renderHepText returns when error and error text", () => {
  const wrapper = shallow(generateComponent({
    errorText: "error",
  }));
  wrapper.setState({ error: true });
  const result = wrapper.instance().renderHelpText();
  expect(result).toMatchSnapshot();
});

it("renderHepText returns when status", () => {
  const wrapper = shallow(generateComponent());
  wrapper.setState({ status: "some status" });
  const result = wrapper.instance().renderHelpText();
  expect(result).toMatchSnapshot();
});

it("style returns blank object whe no style and not disabled", () => {
  const wrapper = shallow(generateComponent({
    style: null,
  }));
  const result = wrapper.instance().style();
  expect(result).toEqual({});
});

it("style returns styles in props", () => {
  const style = {
    color: "red",
  };
  const wrapper = shallow(generateComponent({
    style,
  }));
  const result = wrapper.instance().style();
  expect(result).toEqual(style);
});

it("style returns styles in props plus disabled state", () => {
  const style = {
    color: "red",
  };
  const wrapper = shallow(generateComponent({
    style,
    disabled: true,
  }));
  const result = wrapper.instance().style();
  expect(result).toEqual({
    ...style,
    ...{
      cursor: "inherit",
    },
  });
});
