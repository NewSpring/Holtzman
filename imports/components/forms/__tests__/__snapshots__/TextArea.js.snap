exports[`test appends extra classes 1`] = `
<div className="input append me" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test can hide the label 1`] = `
<div className="input" style={{...}}>
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test has active state 1`] = `
<div className="input input--active" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test has error state 1`] = `
<div className="input input--alert" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test has focused state 1`] = `
<div className="input input--focused" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test overrides with theme if present 1`] = `
<div className="override" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;

exports[`test renderHepText returns when error and error text 1`] = `
<span
  className="input__status">
  error
</span>
`;

exports[`test renderHepText returns when status 1`] = `
<span
  className="input__status">
  some status
</span>
`;

exports[`test renders with props 1`] = `
<div className="input" style={{...}}>
<Label labelFor="2" labelName="myLabel" disabed={[undefined]} />
<textarea id="2" type={[undefined]} placeholder="myPlaceholder" name="myName" className="input classes" disabled={[undefined]} onBlur={[Function]} onFocus={[Function]} onChange={[Function]} style={{...}} rows={10} defaultValue={{...}} />
</div>
`;
